### ISSUE TITLE
[BUG] when using nginx in front of multiple bokeh server instances, an app may be loaded multiple times

### ISSUE THREAD

**Issue by chris-ratcliffe** [bokeh/bokeh#9231]
### FOLLOW THESE INSTRUCTIONS CAREFULLY

*ISSUES THAT DO NOT CONTAIN NECESSARY INFORMATION MAY BE CLOSED, IMMEDIATELY*

The issue tracker is NOT the place for general support. For questions and technical assistance, come join the [Bokeh  Project Discourse](https://discourse.bokeh.org).

For defects or deficiencies, please provide *ALL OF THE FOLLOWING*:

#### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)
bokeh server version: 1.1.0
python version: 3.6.6

#### Description of expected behavior and the observed behavior
I have deployed a number of jpython notebooks that I load through a standalone bokeh server application.  I run multiple instance of the bokeh server with nginx in front.  Frequently, when I load one of the notebooks via an http call, the python code runs more than once.  The expectation is that it will only run once.  Running twice adds a performance hit, even though it eventually does work.

After discussion with @bryevdv at https://discourse.bokeh.org/t/using-standalone-bokeh-server-to-display-jupyter-notebooks-that-contains-some-holoview-objects/4015/6 , it seems this is a known issue.  

#### Complete, minimal, self-contained example code that reproduces the issue

```
# code goes here between backticks

```

**Comment by bryevdv** [comment]
This has been on my mind recently so it's good to have an issue. As far as I can think of, there are no downsides to just not doing any session creation on the HTTP request, and always leaving it to happen at the websocket connect.

**Comment by bryevdv** [comment]
Well, unfortunately there is a reason, and I am not sure what we could easily do about it. The HTTP request handler is responsible for rendering the page HTML. That includes the script to make the WS connection, but also the page title and any user-supplied template. Those are gotten here:
```
title=session.document.title,
template=session.document.template,
template_variables=session.document.template_variables
```
i.e. the HTTP requests handler requires a session be created (which means requires the app code to be run) in order to create the initial page. I am not sure offhand how this could be worked around. The templates in particular are Jinja templates, which means they have to be evaluated on the Python side (so we could not e.g. just push all the work on to the JS side after the WS connection is made)

The only thing I can think of is allowing users to supply another sidecar module of some sort that provides these values, and if the extra module is given then the it is used instead of evaluating the code and creating a real session. That's stating to get quite complicated and ugly.

**Comment by bryevdv** [comment]
@chris-ratcliffe  can you attempt basic session persistence as described here:

https://docs.nginx.com/nginx/deployment-guides/load-balance-third-party/apache-tomcat/#session-persistence-basic

If  that works I think the best option will simply be to document it (and the need for it) in order to avoid this situation when running Bokeh serve behind a load balancer.

**Comment by chris-ratcliffe** [comment]
@bryevdv , currently, I don't have nginx setup as a load balancer.  I am only using it for reverse proxy and ssl termination.  The bokeh server itself is using the --num-procs 0 option to fork multiple tornado servers under the covers.  This is intended to provide scale for my solution, but perhaps it isn't doing what I thought it was doing.    

The solution you provided in the link above requires the use of the upstream stanza, which implies setting up nginx as an explicit load balancer, and running multiple bokeh servers.  I can certainly try setting it up like that.  In that case, does it still make sense to have multiple tornado servers under each bokeh server or should I just use --num-procs 1?

**Comment by chris-ratcliffe** [comment]
After setting up explicit load balancing between nginx/bokeh, and setting num_procs to 1, I no longer see the double load.  However, if I turn num_procs back on, so that there are multiple tornado servers again, I do see the double load happening again.  It would appear that the persistence doesn't percolate down to the bokeh/tornado level.  This is fine, I will just handle my load balancing/scaling at the nginx/bokeh level instead, but I just wanted to let you know.

**Comment by bryevdv** [comment]
@chris-ratcliffe Thank you for running the experiment. Unless Tornado itself offers some solution for sticky sessions with `--num-procs` then there is nothing that can be done in that case. I will look in to that, but at a minimum will use this issue to clearly document the state of affairs and what can be expected in various configurations.

**Comment by chris-ratcliffe** [comment]
@bryevdv , I can confirm that the use of the sticky sessions in conjunction with using nginx as a load balancer in front of multiple bokeh processes prevents the double load issue I was originally seeing, as long as only a single tornado server is used under each bokeh server, ie. --num_procs 1.

**Comment by bryevdv** [comment]
@chris-ratcliffe just to clarify, if I understand correctly, load balancing with nginx and --num-procs=1 works regardless of sticky sessions?

**Comment by chris-ratcliffe** [comment]
Well, I have sticky sessions enabled:

upstream myapp {
      hash $remote_addr;           # Enable simple sticky sessions based on ip

but now that I think about it, I implemented all of the changes at the same time, which is to say, setting up the load balancer and multiple bokeh server processes, and reducing to a single tornado server per bokeh server.  Its possible that I didn't need stick sessions at all, and simply reducing the number of tornado servers down to 1 would have been sufficient.  I will try to find some time to test that later today.

**Comment by chris-ratcliffe** [comment]
@bryevdv , I turned sticky sessions off, and actually, as you suggested, its still ok.  So the ultimate issue seems to have been the --num_procs value, and starting multiple tornado servers.  I think that is the layer where the sticky sessions would be needed to avoid the double load issue.  

Converting my horizontal scaling from the use of multiple tornado servers with one bokeh server, to multiple bokeh servers fronted by nginx load balancing is essentially a workaround to the issue, but one that is fine for me.

**Comment by bryevdv** [comment]
> I think that is the layer where the sticky sessions would be needed to avoid the double load issue.

Agree, and I was hoping the approach to sticky sessions in the link above would be sufficient to allow things to work better when `--num-procs` is used, but evidently not. AFAIK `--num-procs` does not work on Windows at all, so I am starting to think maybe we should consider just getting rid of it for Bokeh 2.0. Or at least making it throw up a few paragraphs of disclaimers and info about limitations. 

cc @bokeh/dev thoughts?

**Comment by bryevdv** [comment]
Just thinking out loud but another possible approach is to set some sort of testable condition that could be used to distinguish between "running in the HTTP request" vs "really running in the WS handler" Then app code could skip anything that does real work in the HTTP request run... that would be pretty ugly but at least provide an avenue for when optimizing this is important.
